feat: Page Types Milestone 1

bapmrl opened the pull request:
## Context

<!--
Please include either
- a ticket
    - [Link text Here](https://link-url-here.org)

- a github issue / forum thread
    - Fixes #

- a detailed description of the issue and it's implications

This part should always include acceptance criteria weither they are accessible on a link or written here directly.
-->




## The Solution

<!--
Highlight explanations where the complexity in your development is.
Keep in mind that the reviewer might not have as much context as you do.
This is very important to make sure the review is thorough and the reviewer understand your changes.
-->




## Impact / Dependencies

<!--
List all the impacts your development might have on internal and external features.
Ideally this should have been discussed prior to this development.

Link of any other PRs that are related to this development.
They should also be referenced in the ticket for reviews.
-->




## Checklist before requesting a review
- [ ] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [ ] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.




## [OPT] Preview

<!--
Visual aid of your work.
It can either be screenshots or a video.
This could help reviewers to get context quickly.
-->




<!--
A funny animal picture is welcome to close your PR!
You can find one here https://unsplash.com/s/photos/funny-animal-picture
-->

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.0
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.0
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.0
 - @slicemachine/init@2.0.2-dev-page-types.0
 - @slicemachine/manager@0.2.2-dev-page-types.0
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.0
 - slice-machine-ui@1.0.3-dev-page-types.0
 - start-slicemachine@0.7.3-dev-page-types.0

xrutayisire committed:
feat: remove onboarding screen (#963)

xrutayisire committed:
chore: update package-lock.json

xrutayisire committed:
ci: skip e2e to fix slice-machine-ui package

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.1
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.1
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.1
 - @slicemachine/init@2.0.2-dev-page-types.1
 - @slicemachine/manager@0.2.2-dev-page-types.1
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.1
 - slice-machine-ui@1.0.3-dev-page-types.1
 - start-slicemachine@0.7.3-dev-page-types.1

xrutayisire committed:
ci: restore e2e job

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.2
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.2
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.2
 - @slicemachine/init@2.0.2-dev-page-types.2
 - @slicemachine/manager@0.2.2-dev-page-types.2
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.2
 - slice-machine-ui@1.0.3-dev-page-types.2
 - start-slicemachine@0.7.3-dev-page-types.2

bapmrl committed:
chore: write components using @prismicio/editor-ui (#968)

bapmrl committed:
Merge branch 'master' into dev-page-types

bapmrl committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.3
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.3
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.3
 - @slicemachine/init@2.0.2-dev-page-types.3
 - @slicemachine/manager@0.2.2-dev-page-types.3
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.3
 - slice-machine-ui@1.0.3-dev-page-types.3
 - start-slicemachine@0.7.3-dev-page-types.3

xrutayisire committed:
feat(page types): update Custom Types creation and edition with default values and restrictions (#974)

MarcMcIntosh committed:
page-types: add seo meta data to custom-type

MarcMcIntosh committed:
Revert "page-types: add seo meta data to custom-type"

This reverts commit d6628b357bef31f2d844cda6c2372e6b37bc65b9.

xrutayisire committed:
feat(slice-machine-ui): New table component (#978)

Co-authored-by: Baptiste Morelle

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.4
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.4
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.4
 - @slicemachine/init@2.0.2-dev-page-types.4
 - @slicemachine/manager@0.2.2-dev-page-types.4
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.4
 - slice-machine-ui@1.0.3-dev-page-types.4
 - start-slicemachine@0.7.3-dev-page-types.4

xrutayisire committed:
feat(page types): new custom types page (#983)

hypervillain committed:
Track page format (#981)

Co-authored-by: Xavier Rutayisire <xavier.rutayisire@gmail.com>

MarcMcIntosh committed:
Page types default fields (#982)

Co-authored-by: Xavier Rutayisire <xavier.rutayisire@gmail.com>

xrutayisire committed:
New Page types Next.js Page (#984)

Co-authored-by: MarcMcIntosh <marcmcintosh1987@gmail.com>

xrutayisire committed:
Improve wording & fix tests

xrutayisire committed:
Restore prismic-types metadata

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.5
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.5
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.5
 - @slicemachine/init@2.0.2-dev-page-types.5
 - @slicemachine/manager@0.2.2-dev-page-types.5
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.5
 - slice-machine-ui@1.0.3-dev-page-types.5
 - start-slicemachine@0.7.3-dev-page-types.5

xrutayisire committed:
Fix pages display for Custom/Page types

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.6
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.6
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.6
 - @slicemachine/init@2.0.2-dev-page-types.6
 - @slicemachine/manager@0.2.2-dev-page-types.6
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.6
 - slice-machine-ui@1.0.3-dev-page-types.6
 - start-slicemachine@0.7.3-dev-page-types.6

xrutayisire committed:
feat(page types): Let a user create/delete a slice zone on Custom/Page Types

xrutayisire committed:
feat(page types): Use custom type config for match url config

xrutayisire committed:
feat(page types): Handle space correctly for Page/Custom types message

xrutayisire committed:
feat(page types): Slice Zone should be here by default for all Page Types

xrutayisire committed:
feat(page types): Handle UID removable if Custom Type is Single

xrutayisire committed:
fix(slice-machine-ui): Prevent Tabs to stay sticky on top of content

xrutayisire committed:
feat(page types): Correctly handle default SEO metadata image field

xrutayisire committed:
feat(page types): Correctly handle URLs redirection and match

xrutayisire committed:
feat(page types): Fix test for default SEO metadata image field

xrutayisire committed:
Merge branch 'master' into dev-page-types

xrutayisire commented on packages/slice-machine/components/Simulator/index.tsx:
```diff
@@ -307,3 +304,11 @@ const Simulator: ComponentWithSliceProps = ({ slice, variation }) => {
 
 export default Simulator;
+
+// TODO: change this config depending on the environment.
```
Should we do this before merging?

bapmrl replied:
Already tracked in https://linear.app/prismic/issue/DT-1333/[simulator]-sm-uses-staging-unsplashwroomio-endpoint.

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.7
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.7
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.7
 - @slicemachine/init@2.0.2-dev-page-types.7
 - @slicemachine/manager@0.2.2-dev-page-types.7
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.7
 - slice-machine-ui@1.0.3-dev-page-types.7
 - start-slicemachine@0.7.3-dev-page-types.7

bapmrl commented on packages/manager/src/managers/customTypes/CustomTypesManager.ts:
```diff
@@ -24,12 +24,17 @@ import { SLICE_MACHINE_USER_AGENT } from "../../constants/SLICE_MACHINE_USER_AGE
 };
 
+type CustomTypesManagerReadAllCustomTypesArgs = {
+	format?: CustomTypeFormat;
```
❓ Should we make `format` mandatory if the whole `args` is optional?

bapmrl commented on packages/slice-machine/components/AppLayout/Navigation/index.tsx:
```diff
@@ -4,10 +4,14 @@ import useWindowSize from "src/hooks/useWindowSize";
 import { useNetwork } from "@src/hooks/useNetwork";
 import { useUnSyncChanges } from "@src/hooks/useUnSyncChanges";
+import { DatabaseIcon } from "@src/components/Icons/DatabaseIcon";
+import { SliceListIcon } from "@src/components/Icons/SliceListIcon";
```
💡 I would tend to use icon names that are more generic and come from Materials Symbols/Icons, as it's what's done in the Editor's repository. One of those is certainly `FolderIcon` and the other I couldn't find... maybe it could be `ViewCarousel`, but I'm not sure.

bapmrl commented on packages/slice-machine/src/features/customTypes/customTypesConfig.ts:
```diff
@@ -0,0 +1,25 @@
+  custom: {
+    title: "Custom Types",
+    name: "Custom Type",
+    urlPathSegment: "custom-types",
```
💡 I've noticed we're duplicating several times the same code that uses these `urlPathSegment` and `urlDynamicSegment` config. What do you think of providing functions like these:
```
export const CUSTOM_TYPES_CONFIG = {
  custom: {
    builderPageDynamicSegment: "customTypeId",
    getBuilderPagePathname: (customTypeId: string) =>
      `/custom-types/${customTypeId}`,
    matchesTablePagePathname: (pathname: string) =>
      pathname === CUSTOM_TYPES_CONFIG.custom.tablePagePathname,
    tablePagePathname: "/custom-types",
  },
  page: {
    builderPageDynamicSegment: "pageTypeId",
    getBuilderPagePathname: (pageTypeId: string) => `/page-types/${pageTypeId}`,
    matchesTablePagePathname: (pathname: string) =>
      pathname === "/" ||
      pathname === CUSTOM_TYPES_CONFIG.page.tablePagePathname,
    tablePagePathname: "/page-types",
  },
};
```

xrutayisire committed:
feat(page types): Update max value for Input placeholder field to 100 to accept new SEO & Metadata default fields placeholder

xrutayisire committed:
feat(page types): Use doc new style change from Custom Type to Custom type

bapmrl commented on packages/slice-machine/components/DeleteCTModal/index.tsx:
```diff
@@ -10,18 +10,18 @@ import { MdOutlineDelete } from "react-icons/md";
-  LocalOnlyCustomType,
-} from "@lib/models/common/ModelData";
+import { CustomType } from "@prismicio/types-internal/lib/customtypes";
+import { CustomTypeFormat } from "@slicemachine/manager/*";
```
⚠️ That's a weird import path.

xrutayisire committed:
feat(page types): Bump versions

xrutayisire committed:
feat(page types): Use doc new style for lowercase custom type in the middle of a sentence

xrutayisire committed:
Publish

 - @slicemachine/adapter-next@0.1.3-dev-page-types.8
 - @slicemachine/adapter-nuxt@0.1.3-dev-page-types.8
 - @slicemachine/adapter-nuxt2@0.1.3-dev-page-types.8
 - @slicemachine/init@2.0.2-dev-page-types.8
 - @slicemachine/manager@0.2.2-dev-page-types.8
 - @slicemachine/plugin-kit@0.2.1-dev-page-types.8
 - slice-machine-ui@1.0.3-dev-page-types.8
 - start-slicemachine@0.7.3-dev-page-types.8

bapmrl commented on packages/slice-machine/components/Forms/components/SelectRepeatable.tsx:
```diff
@@ -19,9 +28,9 @@ export const SelectRepeatable: React.FC = () => {
           }}
         >
-          Repeatable type
+          Reusable type
```
❓ I'm lost, should it be "reusable" or "repeatable"? If uncorrect, we should rename the corresponding icons.

bapmrl replied:
=> We display "reusable", but it's still "repeatable" in the types/code.

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/Layout/Header/index.tsx:
```diff
@@ -29,17 +31,28 @@ const CustomTypeHeader = () => {
             <>
-              <MdSpaceDashboard />
-              <Text>Custom Types</Text>
+              {currentCustomType.format === "custom" ? (
```
⚠️ Icon should probably be in the config. We should check if we have other conditions based on the `format`. It's better to avoid them to ease adding new formats in the future.

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/DeleteSliceZoneModal.tsx:
```diff
@@ -0,0 +1,98 @@
+            }}
+          >
+            <Heading sx={{ fontSize: "14px", fontWeight: "bold", ml: 1 }}>
+              Do you really want to delete Slice Zone?
```
⚠️ "Do you really want to delete the slice zone?".

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/DeleteSliceZoneModal.tsx:
```diff
@@ -0,0 +1,98 @@
+          </Flex>
+        )}
+      >
+        <Paragraph>You're about to delete Slice Zone.</Paragraph>
```
⚠️ "You're about to delete the slice zone.".

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/index.tsx:
```diff
@@ -73,36 +74,41 @@ const mapAvailableAndSharedSlices = (
         : { availableSlices: [], slicesInSliceZone: [], notFound: [] },
     [sliceZone, libraries]
   );
+  const [isSliceZoneActive, setIsSliceZoneActive] = useState(!!sliceZone);
```
⚠️ I don't understand why we need a React state here. Why not just use `const isSliceZoneActive = !!sliceZone`? Are we sure the local React state will always be in sync with the one in Redux?

bapmrl commented on packages/slice-machine/lib/builders/SliceBuilder/FieldZones/index.tsx:
```diff
@@ -140,6 +140,7 @@ const FieldZones: React.FunctionComponent<FieldZonesProps> = ({
           `slice.primary${transformKeyAccessor(key)}`
         }
         dataCy="slice-non-repeatable-zone"
+        isRepeatableCustomType={undefined}
```
❓ Why `undefined`? It's a bit confusing as the name of the prop suggests a `boolean`.

bapmrl commented on packages/slice-machine/lib/models/common/CustomType/index.ts:
```diff
@@ -8,11 +8,17 @@ import { SlicesSM } from "../Slices";
 
 export * from "./tab";
 
+export const CustomTypeFormat = {
```
❓ Do we need to export it?

bapmrl commented on packages/slice-machine/src/components/Breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,10 @@
+  children: ReactNode;
+}>;
+
+export const Breadcrumb: FC<BreadcrumbProps> = ({ children }) => {
```
⚠️ Let's add an `index.ts` file for consistency (when `import`ing this component we don't want `Breadcrumb` to appear twice).

bapmrl commented on packages/slice-machine/src/components/Icons/DatabaseIcon.tsx:
```diff
@@ -0,0 +1,18 @@
+import { FC } from "react";
+
+export const DatabaseIcon: FC = () => (
```
💡 I wonder if we should put icons in a `src/icons` folder rather than nesting them in `src/components/Icons`. `Icons` (capitalized) makes me think I could find an `index.ts` file inside it, but it's not the case.

bapmrl commented on packages/slice-machine/src/features/customTypes/customTypesConfig.ts:
```diff
@@ -0,0 +1,31 @@
+    urlPathSegment: "page-types",
+    urlDynamicSegment: "pageTypeId",
+    blankSlateImage: "/blank-slate-page-types.png",
+    hintSingle: "e.g. home, privacy policy, sign up",
```
💡 I think this module could gain clarity (in terms of responsibility) by moving messages to a separate `customTypesMessages` file.

bapmrl commented on packages/slice-machine/src/modules/availableCustomTypes/factory.ts:
```diff
@@ -1,18 +1,108 @@
+    : tabWithUID;
+}
+
+export const createCustomType = (
```
⚠️ We should probably move this to `src/features/customTypes/customTypesTable` (if we can create a CT only from the table page). That way, we're slowly removing code from the Redux specific `modules` folder.

bapmrl commented on packages/slice-machine/src/modules/useSliceMachineActions.ts:
```diff
@@ -74,6 +74,7 @@ import type {
 import { saveSliceMockCreator } from "./simulator";
 import { SaveSliceMockRequest } from "@src/apiClient";
 import { VariationSM, WidgetsArea } from "@lib/models/common/Slice";
+import { CustomTypeFormat } from "@slicemachine/manager/*";
```
⚠️ Weird import.

bapmrl commented on packages/slice-machine/test/components/Table.test.tsx:
```diff
@@ -0,0 +1,86 @@
+} from "@src/components/Table";
+import { render, screen } from "test/__testutils__";
+
+vi.mock("next/router", () => require("next-router-mock"));
```
❓ Why mock `next/router`?

bapmrl commented on packages/slice-machine/test/components/Table.test.tsx:
```diff
@@ -0,0 +1,86 @@
+  };
+};
+
+describe("Table", () => {
```
❓ Why not have these tests in the same folder as the `Table` component?

bapmrl commented on packages/slice-machine/test/pages/custom-types.test.tsx:
```diff
@@ -313,7 +323,8 @@ describe("Custom Type Builder", () => {
   });
 
-  test("it should send a tracking event when the user saves a custom-type", async (ctx) => {
+  // FIXME: events happening out of order
```
⚠️ Is this important for release?

bapmrl commented on packages/start-slicemachine/src/lib/createSliceMachineExpressApp.ts:
```diff
@@ -105,16 +105,24 @@ export const createSliceMachineExpressApp = async (
+			);
+		});
+
+		app.get("/custom-types", (_req, res) => {
```
❓ Should we have have one for `/page-types`? We should check what Next compiles.

bapmrl commented on packages/start-slicemachine/package.json:
```diff
@@ -55,6 +55,7 @@
 		"serve-static": "^1.15.0"
 	},
 	"devDependencies": {
+		"@prismicio/types-internal": "2.0.0-beta.1",
```
❓ Why are we adding this?

bapmrl commented on CONTRIBUTING.md:
````diff
@@ -38,7 +38,7 @@ cd e2e-projects/next
 ```
 
-Now you can go to your browser and open `localhost:9999` to see the Slice Machine app up and running with some demo Slices and Custom Types.
+Now you can go to your browser and open `localhost:9999` to see the Slice Machine app up and running with some demo Slices and types.
````
⚠️ That looks weird with "Slices".
